api/app/models/slot.scala

```diff
@@ -10,10 +23,169 @@ final case class Slot(
 
 object Slot {
   implicit val testFmt = Json.format[Slot]
+
+  def read(json: String): Option[Slot] = {
+    Json.fromJson[Slot](Json.parse(json)) match {
+      case JsSuccess(slot, _) => Some(slot)
+      case JsError(errors) =>
+        println("unable to deserialise slot from JSON", errors, json)
+        None
+    }
+  }
+}
+
+trait SlotStore {
+  def getAll(): Future[List[Slot]]
+  def get(id: String): Future[Option[Slot]]
+  def upsert(slotId: String, updatedSlot: Slot): Future[Unit] // depends on insert case handling
+
+  // DEV method to ensure index exists and has some dummy data
+  def populate(): Future[Unit]
+}
+
+object SlotStore {
+  def apply(lifecycle: ApplicationLifecycle)(implicit ec: ExecutionContext): SlotStore = {
+    val useRemoteStore = sys.env.get("USE_REMOTE_STORE").contains("true")
+
+    try {
+      val store = if (true) {
+        ElasticsearchStore(lifecycle)
+      } else {
+        MemoryStore()
+      }
+
+      store.populate()
+      store
+    } catch {
+      case e: Throwable =>
+        println("unable to start ES, falling back to MemoryStore", e.getMessage)
+        MemoryStore()
+    }
+  }
+}
+
+class MemoryStore()(implicit ex: ExecutionContext) extends SlotStore {
+  var slots = Map[String,Slot]() // obviously not threadsafe
+
+  def populate(): Future[Unit] = {
+    Future.successful({slots = TestData.slots})
+  }
+
+  def getAll: Future[List[Slot]] = Future.successful(slots.values.toList)
+
+  def get(id: String): Future[Option[Slot]] = {
+    Future.successful(slots.get(id))
+  }
+
+  def upsert(slotId: String, updatedSlot: Slot): Future[Unit] = {
+    slots = slots + (slotId -> updatedSlot)
+    Future.successful(())
+  }
+}
+
+object MemoryStore {
+  def apply()(implicit ec: ExecutionContext) = new MemoryStore()
+}
+
+
+class ElasticsearchStore(client: RestHighLevelClient, index: String = "slots") (
+    implicit ec: ExecutionContext
+) extends SlotStore {
+
+  class Listener[A] extends ActionListener[A] {
+    val p = Promise[A]()
+    def onResponse(response: A): Unit = p.success(response)
+    def onFailure(e: Exception): Unit = p.failure(e)
+    def fut = p.future
+  }
+
+  private[this] def getAsync(request: GetRequest): Future[GetResponse] = {
+    val listener = new Listener[GetResponse]
+    client.getAsync(request, RequestOptions.DEFAULT, listener)
+    listener.fut
```

Comment:
This worked out really nicely!

Reply:
Yeah :) There's almost certainly an even better way to do this to avoid the remaining duplication, but this is relatively succinct still and very easy to understand.